File: C++/CreationalDesignPattern/AdapterDesignPattern.cpp
#include <iostream>
#include <memory>

struct IPaymentProcessor {
    virtual void pay() = 0;
    virtual ~IPaymentProcessor() = default;
};

struct InHousePaymentProcessor : IPaymentProcessor {
    void pay() override {
        std::cout << "In House Application is processing the payment" << std::endl;
        // 100000 lines of code
    }
};

class Razorpay {
public:
    void makePayment() {
        std::cout << "Razorpay Payment 10203013203103" << std::endl;
    }
};

struct RazorPayAdapter : IPaymentProcessor {
    Razorpay razorpay;

    void pay() override {
        razorpay.makePayment();
    }
};

class Application {
    std::unique_ptr<IPaymentProcessor> paymentProcessor;

public:
    explicit Application(std::unique_ptr<IPaymentProcessor> processor)
        : paymentProcessor(std::move(processor)) {}

    void processPayment() {
        paymentProcessor->pay();
    }
};

int main() {
    Application app(std::make_unique<RazorPayAdapter>());  // factory work
    app.processPayment();
    return 0;
}

File: Python/CreationalDesignPattern/AdapterDesignPattern.py
from abc import ABC, abstractmethod


class IPaymentProcessor(ABC):
    @abstractmethod
    def pay(self):
        pass


class InHousePaymentProcessor(IPaymentProcessor):
    def pay(self):
        print("In House Application is processing the payment")
        # 100000 lines of code


class Razorpay:
    def make_payment(self):
        print("Razorpay Payment 10203013203103")


class RazorPayAdapter(IPaymentProcessor):
    def __init__(self):
        self.razorpay = Razorpay()

    def pay(self):
        self.razorpay.make_payment()


class Application:
    def __init__(self, processor: IPaymentProcessor):
        self.payment_processor = processor

    def process_payment(self):
        self.payment_processor.pay()


def main():
    app = Application(RazorPayAdapter())  # factory work
    app.process_payment()


if __name__ == "__main__":
    main()
